### PR TITLE
8261235: C1 compilation fails with assert(res->vreg_number() == index) failed: conversion check

### PR DESCRIPTION
**Pull request by chhagedorn**
The assertion is hit because we run out of virtual registers in the linear scan in C1 and do not handle it. I fixed it by applying the same bailout as in `LIRGenerator::new_register()`. 

There is also a second issue that `LIR_OprDesc::vreg_max` is too big. It is only used in this bailout code. `OprBits::vreg_max` is defined over `OprBits::data_bits` which uses `OprBits::non_data_bits`. But `OprBits::non_data_bits` does not consider `OprBits::pointer_bits` which results in a too large value for `LIR_OprDesc::vreg_max` and the assertion is hit because we don't bail out, yet. This needs to be fixed as well.

Thanks,
Christian

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261235](https://bugs.openjdk.java.net/browse/JDK-8261235): C1 compilation fails with assert(res->vreg_number() == index) failed: conversion check

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 5fd1b911d3dfb7d85fa8fb9aa7126a260fbc317b
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2543/head:pull/2543`
`$ git checkout pull/2543`

### CODE REVIEW

**Comment by eirbjo** [comment]
I tested this branch on my reproducer and it works like a charm:

    776  302       3       org.jaxen.saxpath.base.Verifier::isXMLLetter (6130 bytes)
    [...]
    compilation bailout: out of virtual registers in linear scan
    4717  302       3       org.jaxen.saxpath.base.Verifier::isXMLLetter (6130 bytes)   COMPILE SKIPPED: out of virtual registers in linear scan (retry at different tier)
    4718  334       4       org.jaxen.saxpath.base.Verifier::isXMLLetter (6130 bytes)

Is it expected that the bailout is logged twice?

First, a log line says: "compilation bailout: out of virtual registers in linear scan", then at the end of the next line: "COMPILE SKIPPED: out of virtual registers in linear scan (retry at different tier)" ?

**Comment by chhagedorn** [comment]
Thanks for verifying it!

> Is it expected that the bailout is logged twice?
> 
> First, a log line says: "compilation bailout: out of virtual registers in linear scan", then at the end of the next line: "COMPILE SKIPPED: out of virtual registers in linear scan (retry at different tier)" ?

Yes, that is expected. We first log the bailout alone here:
https://github.com/openjdk/jdk/blob/3882fda83bebf4a8c8100fd59c37f04610491ce7/src/hotspot/share/c1/c1_Compilation.cpp#L620-L627

And then later, we log the entire failed task with the failure reason (in this case the bailout) here:
https://github.com/openjdk/jdk/blob/3882fda83bebf4a8c8100fd59c37f04610491ce7/src/hotspot/share/compiler/compileBroker.cpp#L2347-L2352

File: src/hotspot/share/c1/c1_LIR.hpp [hunk 1]
```diff
@@ -232,7 +232,7 @@ class LIR_OprDesc: public CompilationResourceObj {
     , last_use_bits  = 1
     , is_fpu_stack_offset_bits = 1        // used in assertion checking on x86 for FPU stack slot allocation
     , non_data_bits  = kind_bits + type_bits + size_bits + destroys_bits + last_use_bits +
-                       is_fpu_stack_offset_bits + virtual_bits + is_xmm_bits
+                       is_fpu_stack_offset_bits + virtual_bits + is_xmm_bits + pointer_bits
```
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_LIR.hpp:235]: Would be nice to have this in the same order as the enum values.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by chhagedorn** [comment]
Thank you Tobias for your review!

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review!

**Pull request closed without merging**